iib/workers/tasks/opm_operations.py

```diff
@@ -26,31 +30,206 @@
+    def __init__(self, purpose: str, port: int):
+        """
+        Initialize the PortFileLock object.
 
-    port_stack = list(range(port_start, port_end))
-    random.shuffle(port_stack)
+        :param str purpose: Purpose of the lock
+        :param int port: The port number to be locked
+        """
+        self.purpose = purpose
+        self.port = port
+        self.locked = False
+        self.filename = os.path.join(
+            tempfile.gettempdir(),
+            f'iib_{purpose}_{port}.lock',
+        )
 
-    log.debug('Get random ports from range [%d, %d)', port_start, port_end)
+    def lock_acquire(self):
+        """
+        Create a file representing port lock.
 
-    while port_stack:
-        yield port_stack.pop(0)
+        Before trying to create a a port-lock file, we try to check if the port is free.
+        """
+        log.debug(f"Attempt to lock port {self.port}.")
+
+        if self.locked:
+            err_msg = f"Error: Port {self.port} is already locked"
+            log.debug(err_msg)
+            raise IIBError(err_msg)
+
+        # check if the port is free, opm service is doing the check too,
+        # however this way, we do not have to rely on their error message format
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        try:
+            s.bind(("localhost", self.port))
+        except socket.error:
+            err_msg = f"Port {self.port} is already in use."
+            log.debug(err_msg)
+            raise AddressAlreadyInUse(err_msg)
+        finally:
+            s.close()
+
+        # create file-lock
+        try:
+            os.open(self.filename, os.O_CREAT | os.O_EXCL | os.O_RDWR)
+        except FileExistsError:
+            err_msg = f"Port {self.port} is already locked by other IIB worker."
+            log.debug(err_msg)
+            raise AddressAlreadyInUse(err_msg)
+
+        self.locked = True
+        log.debug(f"Port {self.port} used as {self.purpose} was locked.")
+
+    def unlock(self):
+        """Delete file representing port lock."""
+        if self.locked:
+            os.remove(self.filename)
+            self.locked = False
+            log.debug(f'Port {self.port} used as {self.purpose} was unlocked. ')
+        else:
+            err_msg = f"Attempt to unlock not-locked {self}."
+            log.error(err_msg)
+            raise IIBError(err_msg)
+
+    def __repr__(self):
+        return f"PortFileLock(port: {self.port}, purpose: {self.purpose}, locked: {self.locked})"
+
+
+def port_file_locks_generator(
+    port_stacks: List[List[int]],
+    port_purposes: List[str],
+) -> Generator[List[PortFileLock], None, None]:
+    """
+    Generate PortFileLock from port_stacks and port_purposes.
+
+    Each item in the port_stacks list represents set of ports, from which list of PortFileLocks
+    will be generated.
+
+    :param List[List[int]] port_stacks: List with List of port numbers for one generated item
```

Comment:
nitpick: should we stick to this type of typing or should we use the one we use everywhere?
`:param list(list(int)) port_stacks: `

Reply:
Well, we do not have it standardized - but everywhere else is used just standard python notation